tcp/tcp.c

```diff
@@ -0,0 +1,382 @@
+// Copyright 2016 by BeeGee, see LICENSE.txt
+//
+// Adapted from: github.com/tuanpmt/esp_bridge, Created on: Mar 4, 2015, Author: Minh
+// Adapted from: rest.c, Author: Thorsten von Eicken
+
+#include "esp8266.h"
+#include "c_types.h"
+#include "ip_addr.h"
+#include "tcp.h"
+#include "cmd.h"
+
+#define TCP_DBG
+
+#ifdef TCP_DBG
+#define DBG_TCP(format, ...) os_printf(format, ## __VA_ARGS__)
+#else
+#define DBG_TCP(format, ...) do { } while(0)
+#endif
+
+typedef struct {
+	char			*host;
+	uint32_t		port;
+	ip_addr_t		ip;
+	struct espconn	*pCon;
+	char			*data;
+	uint16_t		data_len;
+	uint16_t		data_sent;
+	uint32_t		resp_cb;
+	uint8_t			conn_num;
+	uint8_t			sock_mode;
+} TcpClient;
+
+
+// Connection pool for TCP socket clients. Attached MCU's just call TCP_setup and this allocates
+// a connection, They never call any 'free' and given that the attached MCU could restart at
+// any time, we cannot really rely on the attached MCU to call 'free' ever, so better do without.
+// Instead, we allocate a fixed pool of connections an round-robin. What this means is that the
+// attached MCU should really use at most as many TCP connections as there are slots in the pool.
+#define MAX_TCP 4
+static TcpClient tcpClient[MAX_TCP];
+static uint8_t tcpNum = 0xff; // index into tcpClient for next slot to allocate
+
+// Any incoming data?
+static void ICACHE_FLASH_ATTR
+tcpclient_recv_cb(void *arg, char *pusrdata, unsigned short length) {
+	struct espconn *pCon = (struct espconn *)arg;
+	TcpClient* client = (TcpClient *)pCon->reverse;
+
+	uint8_t clientNum = client->conn_num;
+	uint8_t cb_type = 1;
+	DBG_TCP("TCP #%d: Received %d bytes: %s\n", client-tcpClient, length, pusrdata);
+	cmdResponseStart(CMD_RESP_CB, client->resp_cb, 4);
+	cmdResponseBody(&cb_type, 1);	
+	cmdResponseBody(&clientNum, 1);
+	cmdResponseBody(&length, 2);
+	cmdResponseBody(pusrdata, length);
+	cmdResponseEnd();
+	
+	if (client->sock_mode != SOCKET_SERVER) { // We don't wait for a response
+		DBG_TCP("TCP #%d: disconnect after receiving\n", client-tcpClient);
+		espconn_disconnect(client->pCon); // disconnect from the server
+	}
+}
+
+// Data is sent
+static void ICACHE_FLASH_ATTR
+tcpclient_sent_cb(void *arg) {
+	struct espconn *pCon = (struct espconn *)arg;
+	TcpClient* client = (TcpClient *)pCon->reverse;
+
+	uint8_t clientNum = client->conn_num;
+	uint8_t cb_type = 0;
+	DBG_TCP("TCP #%d: Sent\n", client-tcpClient);
+	sint16 sentDataLen = client->data_sent;
+	if (client->data_sent != client->data_len) {
+		// we only sent part of the buffer, send the rest
```

Comment:
Fixed in a way that data packages larger than 2800 bytes can be sent (if the uC has enough memory).